### PR TITLE
fix(jsonc): preserve numeric precision for large integers in JSON body

### LINKED ISSUE #5970: [bug]: Hoppscotch truncates precision for long numeric values in the request payload
### Is there an existing issue for this?

- [x] I have searched existing issues and this bug hasn't been reported yet

### Platform

Desktop App

### Browser

Chrome

### Operating System

macOS

### Bug Description

Hoppscotch truncates precision for long numeric values in the request payload, resulting in incorrect parameter values.

Image

Image

### Deployment Type

Self-hosted (on-prem deployment)

### Version

26.2.1

### PR DESCRIPTION
**Pull request by VivekNeer**
Closes #5970

JSON request bodies containing integers larger than `Number.MAX_SAFE_INTEGER` (> `9007199254740991`) were being silently truncated when sent. For example, `{"id": 9007199254740993}` would arrive at the server as `{"id": 9007199254740992}` — last digit flipped with no warning.

## What's changed

**Precision fix (app + CLI)**

* `packages/hoppscotch-common/src/helpers/editor/linting/jsonc.ts` — Instead of `JSON.stringify(node.value)` for number nodes (which already lost precision at JS parse time via IEEE-754), the fix slices the original source string using the AST node's `offset` and `length`, which are always intact. The `originalText` parameter is threaded through all recursive calls in `convertNodeToJSON`.
* `packages/hoppscotch-cli/src/utils/jsonc.ts` — Same fix applied to the CLI's identical copy of the same utility.
* The raw slice is only used when it matches the JSON number grammar. `jsonc-parser` is lenient and emits a number node for forms like `1.`, which would otherwise be emitted verbatim as invalid JSON; those fall back to the normalized `JSON.stringify(node.value)` as before.

**CLI send path**

* `packages/hoppscotch-cli/src/utils/pre-request.ts` — `getFinalBodyFromRequest` was running the cleaned body through `JSON.parse` → `JSON.stringify`, which re-truncated the integer the `jsonc.ts` fix had just preserved. `JSON.parse` is now kept only as a validity check and the cleaned string is returned as-is.

**Tests**

* `packages/hoppscotch-common/src/helpers/__tests__/jsonc.spec.ts` — New file: boundary values, nested objects, arrays, comment/trailing-comma stripping, negative integers, floats, exponent notation, lenient-literal normalization, small-integer regression.
* `packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts` — Mirror of the above for the CLI copy.
* `packages/hoppscotch-cli/src/__tests__/unit/pre-request.spec.ts` — New file exercising the body through `getEffectiveRESTRequest` (the real send path) rather than `stripComments` in isolation: large integers, nesting, comment stripping, env-var substitution, safe-range regression, and invalid JSON → `PARSING_ERROR`. (Placed under `unit/` because `__tests__/functions/` is excluded in `vitest.config.ts` and never runs.)

## Notes to reviewers

The root cause is in the `"number"` case of `convertNodeToJSON`:

```ts
// Before — node.value is already a JS number, precision lost at parse time
case "number":
  return JSON.stringify(node.value)

// After — slice the raw source string, which is always intact
case "number": {
  const raw = originalText.slice(node.offset, node.offset + node.length)
  return JSON_NUMBER_REGEX.test(raw) ? raw : JSON.stringify(node.value)
}

### CODE REVIEW

**Comment by VivekNeer** [comment]
Force-pushed to squash into a single commit and address AI review feedback:

- Fixed test description: `99999999999999999` is 17 digits, not 18
- Added 4 missing CLI tests to match the common-package suite (MAX_SAFE_INTEGER boundary, negative large integer, combined comment+trailing-comma, small integer regression)
- Added floating-point verbatim-preservation test to both suites

**Review by jamesgeorge007 — CHANGES_REQUESTED** [review]
P1 - the `jsonc.ts` change is correct, but the CLI half never takes effect. In `pre-request.ts` (L637-642), the cleaned body is run through `JSON.parse` then `JSON.stringify`, which re-truncates the integer the fix just preserved, so the CLI still sends `9007199254740992` for `...993`. That's the only place the CLI uses `stripComments` outside tests, so the fix never reaches the send path, and the new tests pass only because they exercise `stripComments` in isolation. Could you fix the round-trip, or drop the CLI changes?

P3 - if you drop them, it's worth updating the description so it no longer says both copies are fixed.

**Comment by VivekNeer** [comment]
Thanks for catching that, the JSON.parse → JSON.stringify round-trip in getFinalBodyFromRequest was undoing the fix. Pushed a follow up that keeps JSON.parse as a validity check only and returns the cleaned string as-is, plus tests that exercise the body through getEffectiveRESTRequest rather than stripComments in isolation. Also rebased onto main.

File: packages/hoppscotch-cli/src/utils/pre-request.ts [hunk 1]
```diff
@@ -636,10 +636,13 @@ function getFinalBodyFromRequest(
...
-      // Try to parse the JSON body
+      // Validate the JSON body. The cleaned string itself is returned rather
+      // than a `JSON.parse` -> `JSON.stringify` round-trip, since that would
+      // coerce integers beyond `Number.MAX_SAFE_INTEGER` to IEEE-754 doubles
+      // and silently lose precision that `stripComments` preserved.
       try {
-        const parsedBody = JSON.parse(cleanedBody);
-        return E.right(JSON.stringify(parsedBody));
+        JSON.parse(cleanedBody);
+        return E.right(cleanedBody);
```
[the remark this answers is not in the document]
Comment (VivekNeer) [reply packages/hoppscotch-cli/src/utils/pre-request.ts:647]: Updated in c76f81b

File: packages/hoppscotch-cli/src/utils/pre-request.ts [hunk 2]
```diff
@@ -636,10 +638,13 @@ function getFinalBodyFromRequest(
...
```
[the remark this answers is not in the document]
Comment (VivekNeer) [reply packages/hoppscotch-cli/src/utils/pre-request.ts:647]: Intentionally left as is. The last-key wins dedupe was a side effect of the JSON.parse/JSON.stringify round-trip, not a documented contract, and it silently dropped user input. The CLI now sends exactly what the user wrote  matching the web/desktop app, which already sends stripComments output directly. How duplicate keys are handled is the receiving server's concern.

File: packages/hoppscotch-common/src/helpers/editor/linting/jsonc.ts [hunk 3]
```diff
@@ -47,10 +47,13 @@ function convertNodeToJSON(node: Node): string {
...
       return `[${node.children
-        .map((child) => convertNodeToJSON(child))
+        .map((child) => convertNodeToJSON(child, originalText))
         .join(",")}]`
     case "number":
-      return JSON.stringify(node.value)
+      // Slice the original source text to preserve full numeric precision.
+      // JSON.stringify(node.value) loses precision for integers beyond
+      // Number.MAX_SAFE_INTEGER because jsonc-parser stores them as JS numbers.
+      return originalText.slice(node.offset, node.offset + node.length)
```
[the remark this answers is not in the document]
Comment (VivekNeer) [reply packages/hoppscotch-common/src/helpers/editor/linting/jsonc.ts]: Fixed in 19ddeec
the raw slice is only used when it matches the JSON number grammar, otherwise it falls back to JSON.stringify(node.value). Note 01 wasn't actually affected: jsonc-parser tokenizes it as just 0 (node length 1), so the output was already 0. Only the trailing-dot form (1.) was a real regression.

File: packages/hoppscotch-cli/src/utils/jsonc.ts [hunk 4]
```diff
@@ -28,10 +28,13 @@ function convertNodeToJSON(node: Node): string {
...
       return `[${node.children
-        .map((child) => convertNodeToJSON(child))
+        .map((child) => convertNodeToJSON(child, originalText))
         .join(",")}]`;
     case "number":
-      return JSON.stringify(node.value);
+      // Slice the original source text to preserve full numeric precision.
+      // JSON.stringify(node.value) loses precision for integers beyond
+      // Number.MAX_SAFE_INTEGER because jsonc-parser stores them as JS numbers.
+      return originalText.slice(node.offset, node.offset + node.length);
```
[the remark this answers is not in the document]
Comment (VivekNeer) [reply packages/hoppscotch-cli/src/utils/jsonc.ts]: Same fix as the common copy in 19ddeec, with tests for 1./-1. normalization and exponent notation in both suites. Went with a regex check against the JSON number grammar instead of try/catch JSON.parse same outcome, but explicit about what "canonical" means and no exception in the hot path.

File: packages/hoppscotch-cli/src/utils/pre-request.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (VivekNeer) [reply packages/hoppscotch-cli/src/utils/pre-request.ts:647]: Intentionally left as is. The last-key wins dedupe was a side effect of the JSON.parse/JSON.stringify round-trip, not a documented contract, and it silently dropped user input. The CLI now sends exactly what the user wrote matching the web/desktop app, which already sends stripComments output directly. How duplicate keys are handled is the receiving server's concern.

**Comment by VivekNeer** [comment]
Addressed the review feedback 

- **fef9cba** — `getFinalBodyFromRequest` now keeps `JSON.parse` as a validity check only and returns the cleaned string as-is, so the CLI actually sends `9007199254740993` instead of `...992`. Added unit tests that go through `getEffectiveRESTRequest` (the real send path) rather than `stripComments` in isolation — including env-var substitution and the invalid-JSON → `PARSING_ERROR` case. (Note: `__tests__/functions/` is excluded in `vitest.config.ts`, so these went under `__tests__/unit/` where they actually run.)
- **19ddeec** — cubic flagged that `jsonc-parser` tolerates `1.` and slicing it verbatim would emit invalid JSON. The raw slice is now only used when it matches the JSON number grammar; otherwise it falls back to the normalized value. Large integers are unaffected.
- Left the duplicate-key behavior alone — the old last-key-wins dedupe was a side effect of the round-trip, and the CLI now matches what the web/desktop app already sends.
- Rebased onto `main`.
